sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c

```diff
@@ -0,0 +1,344 @@
+    uint8_t has_cflist = (pkt->size - 1) >= CFLIST_SIZE;
+    gnrc_lorawan_decrypt_join_accept(mac->appskey, ((uint8_t *) pkt->data) + 1,
+                                     has_cflist, out);
+    memcpy(((uint8_t *) pkt->data) + 1, out, pkt->size - 1);
+
+    iolist_t io = { .iol_base = pkt->data, .iol_len = pkt->size - MIC_SIZE,
+                    .iol_next = NULL };
+    le_uint32_t mic;
+    le_uint32_t *expected_mic = (le_uint32_t *) (((uint8_t *) pkt->data) + pkt->size - MIC_SIZE);
+    gnrc_lorawan_calculate_join_mic(&io, mac->appskey, &mic);
+    if (mic.u32 != expected_mic->u32) {
+        DEBUG("gnrc_lorawan_mlme: wrong MIC.\n");
+        status = -EBADMSG;
+        goto out;
+    }
+
+    lorawan_join_accept_t *ja_hdr = (lorawan_join_accept_t *) pkt->data;
+    gnrc_lorawan_generate_session_keys(ja_hdr->app_nonce, mac->mlme.dev_nonce, mac->appskey, mac->nwkskey, mac->appskey);
+
+    le_uint32_t le_nid;
+    le_nid.u32 = 0;
+    memcpy(&le_nid, ja_hdr->net_id, 3);
+    mac->mlme.nid = byteorder_ntohl(byteorder_ltobl(le_nid));
+    /* Copy devaddr */
+    memcpy(&mac->dev_addr, ja_hdr->dev_addr, sizeof(mac->dev_addr));
+
+    mac->dl_settings = ja_hdr->dl_settings;
+
+    /* delay 0 maps to 1 second */
+    mac->rx_delay = ja_hdr->rx_delay ? ja_hdr->rx_delay : 1;
+
+    gnrc_lorawan_process_cflist(mac, out + sizeof(lorawan_join_accept_t) - 1);
+    mac->mlme.activation = MLME_ACTIVATION_OTAA;
+    status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+
+out:
+    gnrc_pktbuf_release(pkt);
+    mlme_confirm_t *mlme_confirm = gnrc_lorawan_mlme_allocate(mac);
+    mlme_confirm->type = MLME_JOIN;
+    mlme_confirm->status = status;
+
+    mac->netdev.event_callback((netdev_t *) mac, NETDEV_EVENT_MLME_CONFIRM);
+}
+
+void gnrc_lorawan_mlme_backoff_expire(gnrc_lorawan_t *mac)
+{
+    uint8_t counter = mac->mlme.backoff_state & 0x1F;
+    uint8_t state = mac->mlme.backoff_state >> 5;
+
+    if (counter == 0) {
+        switch (state) {
+            case GNRC_LORAWAN_BACKOFF_STATE_1:
+                counter = GNRC_LORAWAN_BACKOFF_TIME_1;
+                state = GNRC_LORAWAN_BACKOFF_STATE_2;
+                mac->mlme.backoff_budget = GNRC_LORAWAN_BACKOFF_BUDGET_1;
+                break;
+            case GNRC_LORAWAN_BACKOFF_STATE_2:
+                counter = GNRC_LORAWAN_BACKOFF_TIME_2;
+                state = GNRC_LORAWAN_BACKOFF_STATE_3;
+                mac->mlme.backoff_budget = GNRC_LORAWAN_BACKOFF_BUDGET_2;
+                break;
+            case GNRC_LORAWAN_BACKOFF_STATE_3:
+            default:
+                counter = GNRC_LORAWAN_BACKOFF_TIME_3;
+                mac->mlme.backoff_budget = GNRC_LORAWAN_BACKOFF_BUDGET_3;
+                break;
+        }
+
+    }
+
+    counter--;
+    mac->mlme.backoff_state = state << 5 | (counter & 0x1F);
+    xtimer_set_msg(&mac->mlme.backoff_timer,
+                   GNRC_LORAWAN_BACKOFF_WINDOW_TICK,
+                   &mac->mlme.backoff_msg, thread_getpid());
+
```

Comment:
and again: remove empty line

Reply:
https://github.com/RIOT-OS/RIOT/pull/11022/commits/83f8846bd22f8c78aa72efef0c1ef25451586eb9